### PR TITLE
Fd 474

### PR DESCRIPTION
**Pull request by Emyrk**
Coinbase Cancelling Feature
Allows ability to cancel an output in a coinbase descriptor. Requires 51% of fed servers to trigger

### CODE REVIEW

File: common/identity/identityManager.go [hunk 1]
```diff
@@ -62,170 +94,6 @@ func RandomIdentityManager() *IdentityManager {
 	return im
 }
 
-func (a *IdentityManager) IsSameAs(b *IdentityManager) bool {
```
Comment (Emyrk) [inline common/identity/identityManager.go]: These lines were moved to the bottom of the file, not deleted

File: common/identity/coinbaseCancelManager.go [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+	// Heights of various descriptor cancel proposals. This list will be
+	// sorted, and is used for garbage collecting proposals that have failed
+	ProposalsList []uint32
+
+	// Proposals is all cancel proposals for a given descriptor height
+	//		[descriptorheight][cancel index][id chain]Cancel identity entry
+	Proposals map[uint32]map[uint32]map[[32]byte]identityEntries.NewCoinbaseCancelStruct
+
+	// Boolean indicator if it's been recorded to the admin block. We do not do this more than once
+	AdminBlockRecord map[uint32]map[uint32]bool
```
Comment (factom-clay) [inline common/identity/coinbaseCancelManager.go]: Add explanation of the indexes

File: common/identity/identityManager.go [hunk 3]
```diff
@@ -33,6 +33,17 @@ type IdentityManagerWithoutMutex struct {
...
+	// Not Marshalled
+	// Tracks cancellation of coinbases
+	CancelManager *CoinbaseCancelManager
+
+	// Map of all coinbase outputs that are cancelled.
+	//	The map key is the block height of the DESCRIPTOR
+	//	The list of ints are the indices of the outputs to be
+	//	removed. The keys from the map should be deleted when the
+	//	descriptor+declaration height is hit.
+	CanceledCoinbaseOutputs map[uint32][]uint32
```
Comment (factom-clay) [inline common/identity/identityManager.go]: explain the indexes

File: common/identity/identityManagerEntryBlock.go [hunk 4]
```diff
@@ -520,3 +535,62 @@ func (im *IdentityManager) ApplyNewCoinbaseAddressStruct(ncas *NewCoinbaseAddres
...
+	err := nccs.VerifySignature(id.Keys[0])
+	if err != nil {
+		return false, false, err
+	}
+
+	// Add the cancel to our tallies
+	im.CancelManager.AddCancel(*nccs)
+
+	// Check if we need to update admin block
+	if a != nil {
```
Comment (factom-clay) [inline common/identity/identityManagerEntryBlock.go]: more commentary about syncing disk/network/...

File: common/identityEntries/newCoinbaseCancel.go [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+	FunctionName []byte //"Server Efficiency"
+	//The third is the root identity ChainID.
+	RootIdentityChainID interfaces.IHash
+	//Forth is coinbase descriptor height
+	CoinbaseDescriptorHeight uint32
+	// Fifth is the descriptor index
+	CoinbaseDescriptorIndex uint32
+	//6th is the identity key preimage.
+	PreimageIdentityKey []byte
+	//7th is the signature of the serialized version, text, chainID, new key, and the timestamp.
```
Comment (factom-clay) [inline common/identityEntries/newCoinbaseCancel.go]: not new key it's the coin base stuff

File: state/dbStateManager.go [hunk 6]
```diff
@@ -1162,10 +1162,22 @@ func (list *DBStateList) ProcessBlocks(d *DBState) (progress bool) {
...
+	// Canceling Coinbase Descriptors
+	list.State.IdentityControl.CancelManager.GC(d.DirectoryBlock.GetDatabaseHeight()) // garbage collect
+
 	///////////////////////////////
 	// Cleanup Tasks
 	///////////////////////////////
 	list.State.Commits.Cleanup(list.State)
+	// This usually gets cleaned up when creating the coinbase. If syncing from disk or dbstates, this routine will clean
+	// up any leftover valid cancels.
+	if d.DirectoryBlock.GetDatabaseHeight() > 0 {
```
Comment (factom-clay) [inline state/dbStateManager.go]: fix this

**Pull request merged**